cpp/src/io/csv/csv_reader.cu

```diff
@@ -921,60 +855,114 @@ gdf_error getUncompressedHostData(const char* h_data, size_t num_bytes, const st
+
+  // Trim lines that are outside range, but keep one greater for the end offset
+  if (raw_csv->byte_range_size != 0) {
+    auto it = h_rec_starts.end() - 1;
+    while (it >= h_rec_starts.begin() &&
+           *it > cu_recstart_t(raw_csv->byte_range_size)) {
+      --it;
+    }
+    if ((it + 2) < h_rec_starts.end()) {
+      h_rec_starts.erase(it + 2, h_rec_starts.end());
+    }
+  }
 
-	// Restrict the rows to nrows if nrows is smaller than the remaining number of rows
-	if (raw_csv->nrows >= 0 && (gdf_size_type)raw_csv->nrows < raw_csv->num_records) {
-		raw_csv->num_records = (gdf_size_type)raw_csv->nrows;
-	}
+  // Discard any blank/empty and comment lines
+  if (raw_csv->opts.skipblanklines ||
+      raw_csv->opts.comment != raw_csv->opts.terminator) {
+    const auto match1 = raw_csv->opts.skipblanklines ? raw_csv->opts.terminator
+                                                     : raw_csv->opts.comment;
+    const auto match2 = raw_csv->opts.comment != '\0' ? raw_csv->opts.comment
+                                                      : match1;
+    h_rec_starts.erase(
+        std::remove_if(h_rec_starts.begin(), h_rec_starts.end(),
+                       [&](cu_recstart_t i) {
+                         return (h_uncomp_data[i] == match1 ||
+                                 h_uncomp_data[i] == match2);
+                       }),
+        h_rec_starts.end());
+  }
 
-	// Exclude the rows user chose to skip at the end of the file
-	if (raw_csv->skipfooter != 0) {
-		raw_csv->num_records = gdf_size_type(max(raw_csv->num_records - raw_csv->skipfooter, gdf_size_type{0}));
-		
-	}
-	
-	// Have to at least read the header row
-	if (raw_csv->header_row >= 0 && raw_csv->num_records == 0) 
-		raw_csv->num_records = 1;
-
-	// If specified, header row will always be the first row in the GPU data
-	raw_csv->header_row = min(raw_csv->header_row, 0l);
-
-	const auto start_offset = h_rec_starts[first_row];
-	const auto end_offset = h_rec_starts[first_row + raw_csv->num_records] - 1;
-	raw_csv->num_bytes = end_offset - start_offset + 1;
-	assert(raw_csv->num_bytes <= h_uncomp_size);
-	raw_csv->num_bits = (raw_csv->num_bytes + 63) / 64;
-
-	// Update the record starts to match the device data (skip missing records, fix offset)
-	for (gdf_size_type i = first_row; i <= first_row + raw_csv->num_records; ++i)
-		h_rec_starts[i] -= start_offset;
-	RMM_TRY(RMM_REALLOC(&raw_csv->recStart, sizeof(cu_recstart_t) * (raw_csv->num_records + 1), 0));
-	CUDA_TRY( cudaMemcpy(raw_csv->recStart, h_rec_starts.data() + first_row, 
-		sizeof(cu_recstart_t) * (raw_csv->num_records + 1), cudaMemcpyDefault));
-
-	// Allocate and copy to the GPU
-	RMM_TRY(RMM_ALLOC ((void**)&raw_csv->data, (sizeof(char) * raw_csv->num_bytes), 0));
-	CUDA_TRY(cudaMemcpy(raw_csv->data, h_uncomp_data + start_offset, raw_csv->num_bytes, cudaMemcpyHostToDevice));
+  raw_csv->num_records = h_rec_starts.size();
 
-	return GDF_SUCCESS;
+  // Exclude the rows before the header row
```

Comment:
Header uses row numbers rather than line numbers. That is, it ignores comment lines and blank lines wile skiprows doesn't. We don't know about any comment or blank lines until we do the filtering.